File: .claude/.planning/lang-spike-document-module/fsharp/Money.fs
/// Money — VO em centavos, single-case DU. Espelha `shared/kernel/money.ts`.
/// Regra do domain.md: branded type + smart constructor; imutável; nenhuma exceção cruza a borda.
module Financial.Domain.Money

// Construtor privado ao módulo: só `fromCents` cria. Equivale ao `as Money` do TS — mas sem cast.
type Money = private Money of int64

type MoneyError =
    | Negative
    | Overflow

let zero = Money 0L

/// Smart constructor. Retorna Result — sem `throw` (regra do domínio).
let fromCents (cents: int64) : Result<Money, MoneyError> =
    if cents < 0L then Error Negative else Ok(Money cents)

let cents (Money c) = c

/// Espelha `Money.greaterThan(a, b)`.
let greaterThan (Money a) (Money b) = a > b

let add (Money a) (Money b) : Result<Money, MoneyError> =
    if a > System.Int64.MaxValue - b then Error Overflow else Ok(Money(a + b))

File: .claude/.planning/lang-spike-document-module/rust/money.rs
//! Money — VO em centavos (i64), newtype. Espelha `shared/kernel/money.ts`.
//! Regra do domain.md: branded type + smart constructor; imutável; nenhum panic cruza a borda.

/// Money em centavos. Campo privado: só `from_cents` constrói.
/// Equivale ao `as Money` auditado do TS — mas aqui é garantia do compilador, não convenção.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub struct Money(i64);

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MoneyError {
    Negative,
    Overflow,
}

impl Money {
    pub const ZERO: Money = Money(0);

    /// Smart constructor. Retorna `Result` — sem `throw`/panic (regra do domínio).
    pub fn from_cents(cents: i64) -> Result<Money, MoneyError> {
        if cents < 0 {
            return Err(MoneyError::Negative);
        }
        Ok(Money(cents))
    }

    pub fn cents(self) -> i64 {
        self.0
    }

    /// Espelha `Money.greaterThan(a, b)` do TS.
    pub fn greater_than(self, other: Money) -> bool {
        self.0 > other.0
    }

    pub fn add(self, other: Money) -> Result<Money, MoneyError> {
        self.0
            .checked_add(other.0)
            .map(Money)
            .ok_or(MoneyError::Overflow)
    }
}
